### PR TITLE
[CMake] Publish swift add_library wrapper for single sdk

### PR DESCRIPTION
**Pull request by kateinoigakukun**
`add_swift_target_library` adds targets for multiple SDKs, so it has options that specify target-SDK dependent values like `SWIFT_MODULE_DEPENDS_LINUX`. 
But these kind of options are not scalable, so this PR extracts function that only adds targets for single SDK.

After this change, we can remove `SWIFT_MODULE_DEPENDS_LINUX` options and loop SDKs to call `add_swift_target_library_single_sdk` with each SDK specific values.

I think this would promote [BuildManifesto.md](https://github.com/apple/swift/blob/master/docs/BuildManifesto.md) project.

cc: @compnerd

### CODE REVIEW

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 1]
```diff
@@ -895,11 +895,48 @@ endfunction()
...
+# TARGET_SDKS
+#   The set of SDKs in which this library is included. If empty, the library
+#   is included in all SDKs.
+function(add_swift_target_library name)
+  set(SWIFTLIB_multiple_parameter_options TARGET_SDKS)
+  cmake_parse_arguments(SWIFTLIB
+                        ""
+                        ""
+                        "${SWIFTLIB_multiple_parameter_options}"
+                        ${ARGN})
```
Comment (compnerd) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:914]: Please use variables to make this easier to understand.  Please use an initialism to make the options standout.
```suggestion
  set(options)
  set(single_parameter_options)
  set(multiple_parameter_options TARGET_SDKS)
  cmake_parse_arguments(ASTL
    "${options}"
    "${single_parameter_options}"
    "${multiple_parameter_options}"
    ${ARGN})
```

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 2]
```diff
@@ -895,11 +895,48 @@ endfunction()
...
+  # If target SDKs are not specified, build for all known SDKs.
+  if("${SWIFTLIB_TARGET_SDKS}" STREQUAL "")
+    set(SWIFTLIB_TARGET_SDKS ${SWIFT_SDKS})
+  endif()
+  list_replace(SWIFTLIB_TARGET_SDKS ALL_APPLE_PLATFORMS "${SWIFT_APPLE_PLATFORMS}")
+
+  # If we are building this library for targets, loop through the various
+  # SDKs building the variants of this library.
+  list_intersect(
+      "${SWIFTLIB_TARGET_SDKS}" "${SWIFT_SDKS}" SWIFTLIB_TARGET_SDKS)
```
Comment (compnerd) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:925]: Why is this needed?  Can you not just iterate over `ASTL_TARGET_SDKS`?

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 3]
```diff
@@ -895,11 +895,48 @@ endfunction()
...
+
+  # loop through the SDKs x architectures
+  foreach(sdk ${SWIFTLIB_TARGET_SDKS})
+    add_swift_target_library_single_sdk(${name} SDK ${sdk} ${ARGN})
```
Comment (compnerd) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:929]:
```suggestion
    add_swift_target_library_single_sdk(${name}
      SDK ${sdk}
      ${ARGN})
```
makes it easier to see at a glance what the parameters are

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 4]
```diff
@@ -1106,6 +1142,8 @@ function(add_swift_target_library name)
                         "${SWIFTLIB_multiple_parameter_options}"
                         ${ARGN})
   set(SWIFTLIB_SOURCES ${SWIFTLIB_UNPARSED_ARGUMENTS})
+  # FIXME: This is just for source-code backward compatibility
+  set(sdk ${SWIFTLIB_SDK})
```
Comment (compnerd) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:1146]: I'm not sure I follow.  What is this maintaining compatibility with?  I think that with this change, we should just update the usage and effectively treat this as an internal implementation that shouldn't be used externally.

**Comment by compnerd** [comment]
CC: @drexin

File: stdlib/cmake/modules/AddSwiftStdlib.cmake (hunk 2 above)
Comment (kateinoigakukun) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake:925]: This line filters out unsupported SDK variants from `ASTL_TARGET_SDKS`. This is moved from original `add_swift_target_library`.

Comment (kateinoigakukun) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake:925]: But `add_swift_target_library_single_sdk` also checks that given `SDK` is valid, so this is not needed 👍

File: stdlib/cmake/modules/AddSwiftStdlib.cmake (hunk 4 above)
Comment (kateinoigakukun) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake]: 👍

**Comment by drexin** [comment]
Nice! I don't think I have anything to add to what @compnerd already said.

**Comment by compnerd** [comment]
Thinking a bit more about this - while the change itself seems like it wont break anything, I don't see how it helps the build problem.

You now have two places where the SDK is referenced.  I think it makes more sense for this to go further and extract the SDK specific handling - the eventual state should have no references to the SDK and so this just increases the work that will need to be done to remove the SDK references.

I also don't follow how this is better than the previous state - it has done nothing for the callers nor does it simplify the implementation.

**Comment by kateinoigakukun** [comment]
I intentionally did nothing but extract helper function to avoid big changes. But I agree that this doesn't do anything helpful at this time.

Should I do extract SDK specific handling in this PR? or separated PR to ease reviewing?

**Comment by kateinoigakukun** [comment]
@compnerd I posted a draft PR that extract SDK specific handling from `add_swift_target_library`.
https://github.com/apple/swift/pull/30972
Do you think it's better than the previous state?

**Comment by compnerd** [comment]
I think that you should at least explain that properly in the commit message then.  Right now, your commit message doesn't explain why this change is useful or what you are planning on doing, so it just seems more verbose and not very useful.

**Comment by kateinoigakukun** [comment]
@compnerd I added a description of desired state and plan.
BTW what do you think about fat binary in CMake when invoking cmake for each architectures? Is it out of scope of CMake work and it should be done in build-script-impl?

**Comment by compnerd** [comment]
FAT binaries are MachO specific.  CMake only supports that for C/C++ because clang has special modes for that.  That’s best done outside of CMake.

**Comment by compnerd** [comment]
Okay, with the better commit message, this makes sense.  I don’t think that we want the add_swift_target_library to reference the os/arch/platform and rather sink that further as that should really be CMAKE_SYSTEM_NAME/CMAKE_SYSTEM_ARCH

**Comment by compnerd** [comment]
@swift-ci please test

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 5]
```diff
@@ -1176,29 +1206,23 @@ function(add_swift_target_library name)
...
-  # SDKs building the variants of this library.
-  list_intersect(
-      "${SWIFTLIB_TARGET_SDKS}" "${SWIFT_SDKS}" SWIFTLIB_TARGET_SDKS)
-
-  foreach(sdk ${SWIFTLIB_TARGET_SDKS})
-    if(NOT SWIFT_SDK_${sdk}_ARCHITECTURES)
-      # SWIFT_SDK_${sdk}_ARCHITECTURES is empty, so just continue
-      continue()
+    if(NOT SWIFT_SDK_${SWIFTLIB_SDK}_ARCHITECTURES)
+      # SWIFT_SDK_${SWIFTLIB_SDK}_ARCHITECTURES is empty, so just continue
```
Comment (compnerd) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:1210]: Please update the comment

Comment (kateinoigakukun) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake]: Updated 👍

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test
